onnxruntime/core/graph/contrib_ops/shape_inference_functions.cc

```diff
@@ -114,6 +114,45 @@ void EmbedLayerNormalizationShapeInference(::ONNX_NAMESPACE::InferenceContext& c
   }
 }
 
+void SkipLayerNormalizationShapeInference(::ONNX_NAMESPACE::InferenceContext& ctx) {
+  propagateShapeAndTypeFromFirstInput(ctx);
+
+  auto stash_type = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+  if (ctx.getNumOutputs() > 1) {
+    auto output_type = ctx.getOutputType(1);
+    output_type->mutable_tensor_type()->set_elem_type(static_cast<int32_t>(stash_type));
+  }
+  if (ctx.getNumOutputs() > 2) {
+    auto output_type = ctx.getOutputType(2);
+    output_type->mutable_tensor_type()->set_elem_type(static_cast<int32_t>(stash_type));
+  }
+  if (ctx.getNumOutputs() > 3) {
+    propagateElemTypeFromInputToOutput(ctx, 0, 3);
+  }
+  if (!hasNInputShapes(ctx, 1)) {
```

Comment:
Better change to if (!hasInputShape(ctx, 0)){ return; }

It is because the following code copy shape from input 0.

Reply:
`!hasNInputShapes(ctx, 1)` is a pattern I've seen in other shape inference. The `1` stands for the first `1` inputs, which in this case would be equivalent to `if (!hasInputShape(ctx, 0)){ return; }`

Reply:
You are right. It means the first N inputs has shape.
https://github.com/onnx/onnx/blob/f1c59a5ff7b51c27833a50bc75f5524b65d05530/onnx/defs/shape_inference.h#L314-L321